# HWK 3 -- part 3

# Problem 1

In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F

### a) Make a linear module (call it mod) with the following weight matrix and bias vector:
$$
W = \begin{bmatrix}
1 &1 \\ 2 & 0 \\ -1 & 1
\end{bmatrix}
\qquad
b = \begin{bmatrix}
1  \\ 1 \\ -1 
\end{bmatrix}
$$
(Hint: the weights and bias are going to be initialized randomly, then you will need to change them one by one.)

In [2]:
mod = nn.Linear(2,3,bias=True)
W = [[1,1], [2,0], [-1,1]]
b = [1,1,-1]
for i in range(len(W)):
    for j in range(len(W[0])):
        mod.weight[i][j] = W[i][j]
for i in range(len(b)) : mod.bias[i] = b[i] 
print(mod.weight)
print(mod.bias)

Parameter containing:
tensor([[ 1.,  1.],
        [ 2.,  0.],
        [-1.,  1.]], grad_fn=<CopySlices>)
Parameter containing:
tensor([ 1.,  1., -1.], grad_fn=<CopySlices>)


### b) Make a vector $x = \begin{bmatrix} 2  \\ 1  \end{bmatrix} $  and feed it to the module. Print the output. 
(Hint: make sure to use torch.Tensor with a capital T! If you use torch.tensor you will get a bug. Why?)

In [3]:
x = torch.Tensor([2,1])
y = mod(x)
print(y)

tensor([ 4.,  5., -2.], grad_fn=<ThAddBackward>)


### c) Check with pen and paper that pytorch got the correct answer. (I will collect it, so write it clearly.)

# --------------------------------------

# Problem 2

### a) Make a 3 layer neural net as follow:
$$
\text{Linear layer 1} \to  \text{ReLU} \to \text{Linear layer 2} \to  \text{ReLU} \to \text{Linear layer 3}
$$
### Set the weight matrices of the three linear layers to be 
$$
W_1 
= \begin{bmatrix}
1 &1 \\ 0 & -1 
\end{bmatrix} \qquad \text{,} \qquad
W_2 
= \begin{bmatrix}
-1 &1 \\ 1 & -1 
\end{bmatrix}
\qquad \text{and} \qquad
W_3 
= \begin{bmatrix}
0 &2 \\ 1 & 1 
\end{bmatrix}
$$
### Let's not put any bias. Put a ReLU nonlinearity between the linear layers, but do NOT put any softmax at the end (so the network will output raw scores).

In [4]:
class three_layer_net(nn.Module):

    def __init__(self, input_size, hidden_size1, hidden_size2, output_size):
        super().__init__()
        
        self.layer1 = nn.Linear(input_size, hidden_size1 , bias=False)
        self.layer2 = nn.Linear(hidden_size1, hidden_size2 , bias=False)
        self.layer3 = nn.Linear(hidden_size2, output_size, bias=False)
        
        
    def forward(self, x):
        
        x = self.layer1(x)
        x = F.relu(x)
        x = self.layer2(x)
        x = F.relu(x)     
        return x

### b) Make a vector $x = \begin{bmatrix} 2  \\ 1  \end{bmatrix} $  and feed it to the network. Print the output.

In [5]:
net = three_layer_net(2,2,2,2)

net.layer1.weight[0][0] = 1
net.layer1.weight[0][1] = 1
net.layer1.weight[1][0] = 0
net.layer1.weight[1][1] = -1
        
net.layer2.weight[0][0] = -1
net.layer2.weight[0][1] = 1
net.layer2.weight[1][0] = 1
net.layer2.weight[1][1] = -1   
        
net.layer3.weight[0][0] = 0
net.layer3.weight[0][1] = 2
net.layer3.weight[1][0] = 1
net.layer3.weight[1][1] = 1

x = torch.Tensor([2,1])
y = net(x)
print(y)

tensor([0., 3.], grad_fn=<ReluBackward>)


### c) Check with pen and paper that pytorch got the correct answer. (I will collect it, so write it carefully)

# ---------------------------------------------

# Problem 3

### a) Use Pytorch to compute the softmax of the following vectors:
$$
x = \begin{bmatrix} 1  \\ 2 \\-2  \end{bmatrix} 
$$

In [6]:
x = torch.Tensor([1,2,-2])
p = F.softmax(x, dim=0)
print(p)

tensor([0.2654, 0.7214, 0.0132])


### b) Check with simple python command that pytorch correctly computed the softmax. Follow the steps below (note that you need to import the package "math" in order to have access to the exponential function).

In [7]:
import math

# compute the exponential of the thre entries of the vector:
y1 = math.exp(1)
y2 = math.exp(2)
y3 = math.exp(-2)
print(y1)
print(y2)
print(y3)



2.718281828459045
7.38905609893065
0.1353352832366127


In [8]:
# then divide by the sum and you should recover the values computed by pytorch
sum_y = y1+y2+y3
p1= y1/sum_y
p2= y2/sum_y
p3= y3/sum_y
print(p1)
print(p2)
print(p3)

0.26538792877224193
0.7213991842739688
0.013212886953789417


### c) Create the matrix
$$
x = \begin{bmatrix} 1  & 2 & -2  \\ 1  & 5 & -2  \end{bmatrix} 
$$
### Note that the first row is the vector from part a). Compute the softmax of this matrix in the dimension 1. 

In [9]:
x = torch.Tensor([[1,2,-2], [1,5,-2]])
p = F.softmax(x, dim=1)
print(p)

tensor([[0.2654, 0.7214, 0.0132],
        [0.0180, 0.9811, 0.0009]])


### d) Explain in the cell below why the result you got make sense. In order to write in a cell (rather than puting some code, you need to switch from "code" to "Markdown" in the box right below the Widget and Help)

Since we put dim 1 that means to compute the softmax over the first dimension which are the columns, that means that I want all the colums values to add up to one (for every row). Hence the softmax is produced on tensors [1,2,-2] and [1,5,-2]. The values make sense, for the first tensor, 2 is the biggest value hence it will be bigger after softmax. Similarly for second tensor, we can see that the only difference from the other tensor is the middle value which is changed from 2 -> 5 this makes a huge difference on te softmax result, again since we are exponentiating, the 5 becomes very large taking a value of 0.9811. 

### Now compute the softmax in the dimension 0, and explain why it makes sense (in particular, explain the 50% probabilities in the first and last columns)

In [10]:
x = torch.Tensor([[1,2,-2], [1,5,-2]])
p = F.softmax(x, dim=0)
print(p)

tensor([[0.5000, 0.0474, 0.5000],
        [0.5000, 0.9526, 0.5000]])


Now we are computing the softmax over the dimension 0 which means that for every column we add up the value on each row and we want it to add to one and performe the softmax on those. This means the softmax is applied on the following [1,1], [2,5], [-2,-2]. The first column has two values 1 and 1, the softmax will give 50% on both because these values are identical ( e^1/2e^1 = 1/2) same argument goes for the last column. for the second column, 5 is greater than 2 hence we see a bigger value of 0.9526 allocated to that position.